File: centerOfMass.lua
-- set weight
function setCustomWeight(part, mass)
	local volume = part.Size.X * part.Size.Y * part.Size.Z
	local density = mass / volume
	part.CustomPhysicalProperties = PhysicalProperties.new(density, part.Elasticity, part.Friction)
end

-- main function
function calculateCenterOfMass(folder)
	local totalMass = 0
	local weightedSum = Vector3.new(0, 0, 0)

	-- hovercraft parts' weight & position
	local parts = {
		{name = "arduinoBoard", mass = 118.7, position = Vector3.new(0, 4, -12)},
		{name = "batteryOne", mass = 36, position = Vector3.new(2, 6, -12)},
		{name = "batteryTwo", mass = 36, position = Vector3.new(-2, 6, -12)},
		{name = "hovercraftBase", mass = 0, position = Vector3.new(0, 1, 0)},  -- Assuming default mass for hovercraftBase
		{name = "iMU", mass = 7.4, position = Vector3.new(0, 4.53, -18)},
		{name = "infraredSensor", mass = 3.53, position = Vector3.new(0, 5, 16)},
		{name = "leftUltrasonicSensor", mass = 9.53, position = Vector3.new(15, 6, 7)},
		{name = "rightUltrasonicSensor", mass = 9.53, position = Vector3.new(-15, 6, 7)},
		{name = "thrustFan", mass = 175.36, position = Vector3.new},
		{name = "liftFan", mass = 199.94, position = Vector3.new(0, 3.75, 0)},
		{name = "servoMotor", mass = 44.14, position = Vector3.new}
	}

	-- assign weights to each parts
	for _, partInfo in pairs(parts) do
		local part = folder:FindFirstChild(partInfo.name)
		if part then
			setCustomWeight(part, partInfo.mass)

			-- Update total mass and weighted sum
			totalMass = totalMass + partInfo.mass
			weightedSum = weightedSum + (partInfo.position * partInfo.mass)
		end
	end

	-- Calculate center of mass
	if totalMass > 0 then
		local centerOfMass = weightedSum / totalMass
		return centerOfMass
	else
		return nil
	end
end

--Calculate and print center of mass for the hovercraft
local hovercraftFolder = workspace.Hovercraft
local com = calculateCenterOfMass(hovercraftFolder)
if com then
	print("Center of Mass: " .. tostring(com))  -- Print CG coordinates

	-- create a small visible marker
	local comMarker = Instance.new("Part")
	comMarker.Size = Vector3.new(1, 1, 1) 
	comMarker.Position = com 
	comMarker.Anchored = true 
	comMarker.Material = Enum.Material.Neon 
	comMarker.BrickColor = BrickColor.new("White") 
	comMarker.Parent = workspace 

else
	print("No parts found in the Hovercraft folder.")
end

File: centerOfMass.m
function com = calculateCenterOfMass()
    parts = 
    {
        'arduinoBoard',         118.7,  [0, 5, 4];   
        'batteryOne',           36,     [-3, 12, 6];  
        'batteryTwo',           36,     [3, 12, 6];   
        'hovercraftBase',       0,      [0, 0, 1];    
        'iMU',                  7.4,    [0, 15.5, 4];  
        'infraredSensor',       3.53,   [0, -16, 5];  
        'leftUltrasonicSensor', 9.53,   [-11, -7, 6]; 
        'rightUltrasonicSensor',9.53,   [11, -7, 6];  
        'thrustFan',            175.36, [0, -14, 12]; 
        'liftFan',              199.94, [0, 0, 3.75]; 
        'servoMotor',           44.14,  [0, -14, 4.25]; 
    };

    totalMass = 0;
    weightedSum = [0, 0, 0];

    for i = 1:size(parts, 1)
        mass = parts{i, 2};
        position = parts{i, 3}; 
        totalMass = totalMass + mass;
        weightedSum = weightedSum + mass .* position;
    end

    if totalMass > 0
        com = weightedSum / totalMass;
        fprintf('Center of Mass: [%.2f, %.2f, %.2f]\n', com(1), com(2), com(3));
        figure;
        hold on;
        grid on;
        xlabel('X');
        ylabel('Z');
        zlabel('Y');
        title('Hovercraft Center of Mass Visualization');

        for i = 1:size(parts,1)
            pos = parts{i,3};
            scatter3(pos(1), pos(2), pos(3), 50, 'b', 'filled');
            text(pos(1), pos(2), pos(3), parts{i,1}, 'VerticalAlignment','bottom','HorizontalAlignment','right');
        end

        scatter3(com(1), com(2), com(3), 100, 'r', 'filled');
        text(com(1), com(2), com(3), 'Center of Mass', 'FontWeight','bold', 'Color','r');

        hold off;
    else
        com = [];
        disp('No valid parts with mass found.');
    end
end
